src/test/HST/EnvironmentTests.hs

```diff
@@ -0,0 +1,302 @@
+qName :: String -- ^ The module name the name is qualified by. An empty string
+                --   signalizes an unqualified name.
+      -> String -- ^ The name without its possible qualification. It is assumed
+                --   that this is a regular identifier and not a symbol.
+      -> S.QName a
+qName "" name    = S.UnQual S.NoSrcSpan (S.Ident S.NoSrcSpan name)
+qName modul name = S.Qual S.NoSrcSpan (S.ModuleName S.NoSrcSpan modul)
+  (S.Ident S.NoSrcSpan name)
+
+-- | Creates a module name based on the given name.
+moduleName :: String -> Maybe (S.ModuleName a)
+moduleName ""      = Nothing
+moduleName modName = Just $ S.ModuleName S.NoSrcSpan modName
+
+-------------------------------------------------------------------------------
+-- Test Modules                                                              --
+-------------------------------------------------------------------------------
+-- | Lines of the test module @modA@.
+modA :: [String]
+modA = ["module A where", "data Foo = Bar | Baz"]
+
+-- | Lines of the test module @modB@.
+modB :: [String]
+modB = ["module B where", "data FooB = BarB | Baz"]
+
+-- | Lines of the test module @modC@.
+modC :: [String]
+modC = ["module C where", "data Foo = Bar | BazC"]
+
+-- | Lines of the test module @modD@.
+modD :: [String]
+modD = ["module D where", "data Bar = Foo | Baz"]
+
+-- | Lines of the test module @modUnnamed@.
+modUnnamed :: [String]
+modUnnamed = ["data Foo = BarUnnamed"]
+
+-------------------------------------------------------------------------------
+-- Tests                                                                     --
+-------------------------------------------------------------------------------
+-- | Test group for "HST.Environment".
+testEnvironment :: Spec
+testEnvironment = describe "HST.Environment" $ do
+  context "with unqualified lookups" $ do
+    it "finds a type name in the current module" $ runTest $ do
+      env <- setupTestEnvironment modA [] []
+      let query     = lookupTypeName (qName "" "Bar") env
+          expResult = [(moduleName "A", Just (qName "" "Foo"))]
+      query `typeNameShouldBe` expResult
+    it "finds constructor entries in an imported module" $ runTest $ do
+      env <- setupTestEnvironment [""] [[importDecl "A" False ""]] [modA]
+      let query     = lookupConEntries (qName "" "Foo") env
+          expResult = [ ( moduleName "A"
+                          , (qName "" "Foo", [qName "" "Bar", qName "" "Baz"])
+                          )
+                      ]
+      query `conEntriesShouldBe` expResult
+    it "finds type names for built-in types" $ runTest $ do
+      env <- setupTestEnvironment modA [] []
+      let unitCon   = S.Special S.NoSrcSpan (S.UnitCon S.NoSrcSpan)
+          query     = lookupTypeName unitCon env
+          expResult = [(moduleName "Prelude", Just unitCon)]
+      query `typeNameShouldBe` expResult
+    it "does not search in type names when supposedly given a constructor name"
```

Comment:
I dont understand this example. `Foo` is a type name and not a constructor name.

Reply:
This is on purpose. There is no constructor with the name `Foo`, but a data type. So, when I want to find a constructor with the name `Foo`, then nothing should be found. If the lookup works incorrectly and does find the data type `Foo`, even though I specified that `Foo` should be a constructor, then this test would hopefully find that error.

Reply:
OK.